### PR TITLE
Put an edit file button on pull request files to allow a quick operation

### LINKED ISSUE #23848: Edit File button during PR review
### Feature Description

We love using Gitea and it would be great that if we could edit files during pull request reviews. As a reviewer I go to "Files Changed" tab on a pull request, and it would be great if there is a button called "Edit File" next to "View File" button. Upon clicking on it could be either;
* Convert file view in current page (Files Changed tab) to editable one similar to commenting and it will create a commit with changes upon save
* Show a popup that has editable file in it and that will create a commit with file changes
* Go to existing edit file page in the source branch of the PR where saving the file will create a commit

This would save time on fixing small typos etc encountered rather than going back and forth between people or pulling the entire branch, changing and pushing :)

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by lunny**
Resolve #23848

This PR put an edit file button on pull request files to allow a quick edit for a file. After the edit finished, it will return back to the viewed file position on pull request files tab.

It also use a branch view file link instead of commit link when it's a non-commit pull request files view.

image

### CODE REVIEW

**Comment by silverwind** [comment]
>  It also moved the escape/unescape to the left.

I think it should stay on right (or be eliminated entirely, but that's another topic).

**Comment by lunny** [comment]
> > It also moved the escape/unescape to the left.
> 
> I think it should stay on right (or be eliminated entirely, but that's another topic).

Done.

File: routers/web/repo/pull.go
Comment (wxiaoguang) [inline routers/web/repo/pull.go]: Where is this variable used?

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]: These variable names do not seem right.

`Can` usually means it is from permission checks. But here it is a UI toggle option.

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]: `back_to_pr` + `back_to_diff_hash` is quite fragile. I think it should use a prepared `return_uri`

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]: I think it should add a `setInterval` JS call in frontend to check whether the current PR has been updated, and then show a "Refresh" button just like GitHub (no need to use websocket at the moment, to keep this PR simple)

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]: `http.StatusXxx` but not a magic number. And 302 has been deprecated for long time.

File: templates/repo/diff/box.tmpl
Comment (wxiaoguang) [inline templates/repo/diff/box.tmpl]: I think it's better to put the "Edit" button after the "View" button

**Comment by delvh** [comment]
Hmm… I think this PR depends on #29400:
It sounds much more easy to use to me when we simply display a menu with actions there instead of filling the available space with more and more buttons.

**Comment by delvh** [comment]
That's also what GitHub is doing:
image

File: routers/web/repo/pull.go
Comment (lunny) [reply routers/web/repo/pull.go]: They will be used on `box.tmpl` which is a party of the pull request files tab.

File: routers/web/repo/editor.go
Comment (lunny) [reply routers/web/repo/editor.go]: I would like to control the return_uri's scope. It should not be redirected to other places currently.

File: templates/repo/diff/box.tmpl
Comment (lunny) [reply templates/repo/diff/box.tmpl]: ec6ab8d2d5a8c2b01b7288fd1d7d58998e50de79

File: routers/web/repo/editor.go
Comment (lunny) [reply routers/web/repo/editor.go]: ec6ab8d2d5a8c2b01b7288fd1d7d58998e50de79

File: routers/web/repo/editor.go
Comment (lunny) [reply routers/web/repo/editor.go]: ec6ab8d2d5a8c2b01b7288fd1d7d58998e50de79

File: routers/web/repo/editor.go
Comment (lunny) [reply routers/web/repo/editor.go]: ec6ab8d2d5a8c2b01b7288fd1d7d58998e50de79

File: routers/web/repo/editor.go
Comment (wxiaoguang) [reply routers/web/repo/editor.go]: Why "other places" would happen? There is a function `RedirectToFirst` to handle "redirect_uri".

**Comment by silverwind** [comment]
> Hmm… I think this PR depends on #29400: It sounds much easier to use to me when we simply display a menu with actions such as `(un)escape` there instead of filling the available space with more and more buttons.

It's a bit of a unusual use case for `<overflow-menu>` given that this is not a tab menu, but I think it should work as long as all affected items bear the `.item` class.

File: routers/web/repo/editor.go
Comment (lunny) [reply routers/web/repo/editor.go]: 06fb88b8c95be465b3bf7fec11f8a0c56b4447a3

**Comment by yekanchi** [comment]
Is this somehow related to resolving conflicts?
would this editing help resolve conflicts?

File: routers/web/repo/pull.go
Comment (a1012112796) [inline routers/web/repo/pull.go]: looks it should be disabled for agit pull request always.

Comment (lunny) [reply routers/web/repo/pull.go]: 7cf364eeb50a601b082a151757cf0ac2ac8754ee

File: templates/repo/editor/commit_form.tmpl
Comment (silverwind) [inline templates/repo/editor/commit_form.tmpl]: Indent this block.

File: templates/repo/diff/box.tmpl
Comment (wxiaoguang) [inline templates/repo/diff/box.tmpl]: I think it needs this:  Improve QueryEscape helper function #29768 

Then it doesn't need such strange Escape/Unescape anymore.

Comment (wxiaoguang) [reply templates/repo/diff/box.tmpl]: With  #29768, it could be :

```
redirect_to={{print $.BackToLink "#diff-" $file.NameHash | QueryEscape}}
```

And backend code doesn't need to to double-unescape

Comment (lunny) [reply templates/repo/diff/box.tmpl]: afe9aaf

File: templates/repo/editor/commit_form.tmpl
Comment (lunny) [reply templates/repo/editor/commit_form.tmpl]: I didn't do that because it will make the PR review more easier.

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]:
```suggestion
```

File: routers/web/repo/pull.go
Comment (wxiaoguang) [inline routers/web/repo/pull.go]: I can see other places use "PathEscapeSegments" for "branch name"

File: templates/repo/diff/box.tmpl
Comment (wxiaoguang) [inline templates/repo/diff/box.tmpl]: I think the `hide_create_pr=true&hide_create_new_branch=` could all be removed.

The meaning of them is: "the editor is for a special purpose and shouldn't have other operations, it should for editing only".

The meaning of "redirect_to" is: "the editor is for a special purpose and after edit it should redirect back".

So I think at the moment, checking "IsEditingFileOnly (has RedirectTo)" also means "HideXxx=true" (no other operation)

File: templates/repo/editor/commit_form.tmpl
Comment (silverwind) [reply templates/repo/editor/commit_form.tmpl]: Would still do it, reviewers can use the "hide whitespace" option, I do that often for template changes.

File: templates/repo/diff/box.tmpl
Comment (lunny) [reply templates/repo/diff/box.tmpl]: But as an editor, more parameters will be more flexible than before. Maybe other places can reuse them in future PRs. i.e. Edit file in the commit files view, then it can create new branch or create PR.

File: templates/repo/editor/commit_form.tmpl
Comment (wxiaoguang) [reply templates/repo/editor/commit_form.tmpl]: Agree to indent.

If the new code is just one level like:

```
{{if NewCode}}
<div>old content</div>
{{end}}
```

I think it is acceptable.

But now the new code makes it become 2 levels:

```
{{if NewCode}}
{{if OldCode}}
<div>old content</div>
{{end}}
{{end}}
```

For this case, IMO indention is a must.

File: templates/repo/diff/box.tmpl
Comment (wxiaoguang) [reply templates/repo/diff/box.tmpl]: Then it should have a better "parameters" system at that time, but not keeping adding "hide_xxx"

When there is a real requirement, the design should be like this: `?editor_mode=editing-only`, but not by a lot of `hide_xxx`.

Comment (wxiaoguang) [reply templates/repo/diff/box.tmpl]: I still think using only "redirect_to" is clear enough for hiding the "create PR/branch" operations. 

Because when it should "redirect back", then the editor page shouldn't create PR/branch.

In this case, less is more, we can leave a comment there, and that's why I suggest to use template variable `IsEditingFileOnly = (redirectTo == "")` to show the real behavior.

I do not think it's worth to make the parameters too complex than it should be.

File: routers/web/repo/pull.go
Comment (lunny) [reply routers/web/repo/pull.go]: 9c89597

File: templates/repo/diff/box.tmpl
Comment (wxiaoguang) [reply templates/repo/diff/box.tmpl]: And one more thing, "redirect_to" doesn't seem a good name, because it has been used by the "auth" middlewares, it would cause conflict.

That's why I suggested to use "return_uri" (https://github.com/go-gitea/gitea/pull/29697#discussion_r1518825237 , or "return_to")

Comment (lunny) [reply templates/repo/diff/box.tmpl]: 4bd286e7aec79589d81ae0d97d21c1e515fab2ff

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go:198]: The last question, how is the "return_uri" passed into the form and pass back by the POST request?

I can only see it is in the "GET" request for the editor, then when "submit" the editor form, does the "return_uri" still exist?

Comment (lunny) [reply routers/web/repo/editor.go:198]: From the pull request files tab -> editor GET request, the parameter has been given
When submitted with POST requests, the queries will be sent from the editor page.

**Review by wxiaoguang — APPROVED** [review]

File: routers/private/hook_post_receive.go
Comment (KN4CK3R) [inline routers/private/hook_post_receive.go:115]: The errors are just logged but do not block further processing. Is this intended?

Comment (lunny) [reply routers/private/hook_post_receive.go:115]: Yes. Keeping this PR the same logic as before. Maybe we can have further discussion and change it in future.

File: templates/repo/editor/commit_form.tmpl
Comment (silverwind) [inline templates/repo/editor/commit_form.tmpl]: These `if` are combinable as `{{if and (not .Repository.IsEmpty) .IsEditingFileOnly}}`, right?

File: templates/repo/editor/commit_form.tmpl
Comment (wxiaoguang) [inline templates/repo/editor/commit_form.tmpl]: Literally, it should be `{{if .ReturnURI}}{{.ReturnURI}}`

Comment (wxiaoguang) [reply templates/repo/editor/commit_form.tmpl]:
```suggestion
	<a class="ui button red" href="{{if .ReturnURI}}{{.ReturnURI}}{{else}}{{$.BranchLink}}/{{PathEscapeSegments .TreePath}}{{end}}">{{ctx.Locale.Tr "repo.editor.cancel"}}</a>
```

File: templates/repo/editor/edit.tmpl
Comment (wxiaoguang) [inline templates/repo/editor/edit.tmpl]:
```suggestion
					<span>{{ctx.Locale.Tr "repo.editor.or"}} <a href="{{if .ReturnURI}}{{.ReturnURI}}{{else}}{{$.BranchLink}}{{if not .IsNewFile}}/{{PathEscapeSegments .TreePath}}{{end}}{{end}}">{{ctx.Locale.Tr "repo.editor.cancel_lower"}}</a></span>
```

File: routers/private/hook_post_receive.go
Comment (wxiaoguang) [reply routers/private/hook_post_receive.go:115]: Add a comment and TODO ?

**Comment by wxiaoguang** [comment]
Any progress? Why stale.

**Comment by lunny** [comment]
> Any progress? Why stale.

I think it's ready to be reviewed again. Need one more approval.

**Comment by silverwind** [comment]
Does not compile for me currently:

```
routers/web/repo/editor.go:85:6: ctx.RedirectToFirst undefined (type *"code.gitea.io/gitea/services/context".Context has no field or method RedirectToFirst)
```

**Comment by silverwind** [comment]
I guess I can try making it a dropdown after above error is fixed.

File: routers/web/repo/editor.go
Comment (wxiaoguang) [inline routers/web/repo/editor.go]:
```suggestion
	ctx.RedirectToCurrentSite(
```

**Comment by wxiaoguang** [comment]
> I guess I can try making it a dropdown after above error is fixed.

I will fix the compile error. But please "make it a dropdown" in a new PR. I don't think it should keep bloating a reviewed PR or keep it stale.

**Comment by lunny** [comment]
> I guess I can try making it a dropdown after above error is fixed.

Maybe you can change the mean in that PR together after this PR is merged.

**Comment by wxiaoguang** [comment]
I would initiate the last call. If no objection, merge it as-is with one approval.

File: tests/integration/pull_compare_test.go
Comment (silverwind) [inline tests/integration/pull_compare_test.go]:
```suggestion
	editButtonCount := doc.doc.Find(".diff-file-header-actions a[href*='/_edit/']").Length()
```

Tag name is too specific.

**Review by silverwind — APPROVED** [review]

**Pull request merged**
